### PR TITLE
Add hot restart button.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
- Calls hotRestart service added in https://github.com/flutter/flutter/pull/26669
- Resets the extensionStatesUpdated completer on IsolateExit.
- Revealed a small bug in parsing event json - fix is in this CL.

### CODE REVIEW

File: lib/timeline/timeline.dart [hunk 1]
```diff
@@ -290,6 +292,37 @@ class TimelineScreen extends Screen {
 
     return button;
   }
+
+  // TODO: move this button out of timeline if we decide to make a global button bar.
```
Comment (jacob314) [inline lib/timeline/timeline.dart]: this button seems pretty similar to the hot reload one. Consider making a base class for buttons calling a service extension of this form. I expect this isn't the last one we will add.

File: test/service_manager_test.dart [hunk 2]
```diff
@@ -166,7 +166,9 @@ void main() {
       await serviceManager.performHotReload();
     });
 
-    // TODO(kenzie): add hot restart test case.
+    test('hotRestart', () async {
+      await serviceManager.performHotRestart();
```
Comment (jacob314) [inline test/service_manager_test.dart]: verify that something happened. For example: 
if you use Eval to set a top level field in flutter, after you call hot restart it won't be set.

File: lib/timeline/timeline.dart (hunk 1 above)
Comment (kenzieschmoll) [reply lib/timeline/timeline.dart]: Done.

File: test/service_manager_test.dart (hunk 2 above)
Comment (kenzieschmoll) [reply test/service_manager_test.dart]: Done.

File: lib/utils.dart [hunk 3]
```diff
@@ -115,6 +115,10 @@ String getUserHomeDir() {
 /// value.
 typedef void VoidFunction();
 
+/// A typedef to represent a function taking a single argument and with no
+/// return value.
+typedef void VoidFunctionWithArg(dynamic arg);
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/utils.dart]: Hmm I ran dartfmt and nothing changed.

 I'll do a dartfmt CL over the whole codebase as a follow up CL.

File: lib/ui/ui_utils.dart [hunk 4]
```diff
@@ -87,3 +92,47 @@ class ServiceExtensionButton {
...
+
+  final RegisteredServiceDescription serviceDescription;
+  final VoidFunction action;
+  final VoidFunctionWithArg errorAction;
+  PButton button;
+
+  void _click() async {
+    try {
+      button.disabled = true;
+      action();
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/ui/ui_utils.dart]: done.

File: lib/timeline/timeline.dart [hunk 5]
```diff
@@ -264,7 +264,7 @@ class TimelineScreen extends Screen {
 
   // TODO: move this button out of timeline if we decide to make a global button bar.
   CoreElement createHotReloadButton() {
-    final VoidFunction action = () async {
+    final VoidFuture action = () async {
```
Comment (jacob314) [inline lib/timeline/timeline.dart]: nit. You could omit this type if you wanted.

File: lib/utils.dart [hunk 6]
```diff
@@ -115,6 +115,10 @@ String getUserHomeDir() {
 /// value.
 typedef void VoidFunction();
 
+/// A typedef to represent a function taking no arguments and returning a void
+/// future.
+typedef Future<void> VoidFuture();
```
Comment (jacob314) [inline lib/utils.dart]: nit: VoidAsyncFunction might be clearer

File: lib/utils.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/utils.dart]: updated manually

e.x. `typedef VoidAsyncFunction = Future<void> Function();`

File: lib/utils.dart (hunk 6 above)
Comment (kenzieschmoll) [reply lib/utils.dart]: renamed

File: lib/utils.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (devoncarew) [reply lib/utils.dart]: I believe that https://github.com/flutter/devtools/pull/127 will catch this in the future.

**Pull request merged**